README.txt
$Id$

Basic Info
==========

The ldap_integration module is actually a set of modules. 
Currently, they include
	ldapauth
	ldapgroups
	ldapdata
	

ldapauth.module is the basic module upon which every other ldapXXX module 
depends on. It implements LDAP/AD authentication for Drupal. This module has 
been tested with OpenLDAP, OpenDirectory (Mac OS X) and Active Directory 2003.

ldapgroups.module depends upon ldapauth.module and extends the functionality 
to  integrate LDAP Groups with Drupal Roles

ldapdata.module depends upon the profile and ldapauth module  and extends the 
functionality to enable mapping of other LDAP attributes to Drupal attributes.
Also allows management of ldap attributes via the Drupal profile module.


History
=======
This module was initially developed by pablom (http://drupal.org/user/6936) 
for drupal. Its current maintainers are kreaper (http://drupal.org/user/57158)
and scafmac (http://drupal.org/user/90087)


Limitations
===========

The module group went through a re-design overhaul as of Drupal 5.x when it 
was transferred to the new maintainers. As such, support for previous modules
is extremely limited. Before upgrading this module from < 4.7.x to 5.x, you
must read http://drupal.org/node/92407#comment-178886


Documentation
=============

Documentation is forthcoming. For LDAP administrators, it should be intuitive.
(well, at least that was the thought..)




ChangeLog
2007-08-23 kreaper <http://rajeev.name>

	* ldapauth.module: http://drupal.org/node/169766 
	same html fixed as 164078. Did not actually fix it last time :)
	

2007-08-22 kreaper <http://rajeev.name>

	* ldapauth.module: http://drupal.org/node/142738
	do not generate random password during new ldap user registration

	* ldap_integration/LDAPInterface.php: http://drupal.org/node/164049
	Allow anonymous LDAP connect();

	* ldapauth.module: http://drupal.org/node/164078
	HTML fixes

	* ldap_integration/LDAPInterface.php: http://drupal.org/node/158671
	No longer initializing $ldap->server to "example.com"
	
2007-07-22 kreaper <http://rajeev.name>

	* ldapauth.module: http://drupal.org/node/158671
	contributed code that modifies code flow and also enabled ldapauth_auth() to work with other modules. 

	* ldapdata.module: bug fix for http://drupal.org/node/136068
	http://bugs.mysql.com/bug.php?id=25520 specifies that in Windows, TEXT/BLOB cannot have a default value

2007-07-09 kreaper <http://rajeev.name>

	* ldapgroups.module: renamed the global variable $ldap to $ldapgroups_ldap. 
	Created a new function _ldapgroups_ldap_init() to setup the ldapgroups_ldap object.

	* ldapdata.module: renamed the global variable $ldap to $ldapdata_ldap

	* ldapauth.module: renamed the global variable $ldap to $ldapauth_ldap

ldap_integration/conf.php
<?php
// $Id$

/*
 * Interesting constants that admins would want to mess with
 */

// Admin must define a user able to write users' attributes (to generate new passwords when they are lost)
define(LDAP_WRITER_USER_DN, 'cn=admin,dc=pablobm,dc=com');
define(LDAP_WRITER_USER_PASS, 'secret');

// Admin must define a user able to read users' attributes (to read e-mail addresses when passwords are lost)
define(LDAP_READER_USER_DN, 'cn=admin,dc=pablobm,dc=com');
define(LDAP_READER_USER_PASS, 'secret');

//   Specifies the relation between LDAP attributes and DrupalDB
// user fields
$GLOBALS['ldap_drupal_mappings'] = array(
  // ldap_attribute => drupal field
  // 'cn' => 'name',
  // 'mail' => 'mail',
  'userPassword' => 'pass',
  // 'signature' => 'signature',
  // 'timeZome' => 'timezone'
);

//   Specifies the relation between LDAP groups and Drupal roles,
// as well as which names the latters should take.
$GLOBALS['ldap_group_role_mappings'] = array(
   // LDAP group => Drupal role
   'cn=DeptTST,ou=Departmental,ou=Groups,dc=myorg,dc=mytld' => 'TST',
   'cn=Admins,dc=myorg,dc=mytld' => 'Admin',
   'cn=Admins,dc=pablobm,dc=com' => 'Admin',
   'Clerks' => 'Clerks',
   'simpleSecurityObject' => 'Secured',
   'posixAccount' => 'RMS worshippers'
);

/*
 * Edit this list so that all attributes that you might wish to show
 * are included. You may then enable/disable attributes in the Admin page.
 *
 * The left hand element indicates the LDAP attribute name.
 *
 * The 1st right hand elements in the array determine how the attribute and
 * its value are displayed. The first element may be 'text', 'image', or 'url'.
 */

$GLOBALS['ldap_attributes'] = array(
//  'givenName'     => array('text', 'textfield', t('First Name'), 64, 64),
//  'userPassword'  => array('pass', 'password', t('Password'), 64, 64),
//  'initials'      => array('text', 'textfield', t('Initial'), 64, 4, t('Your middle initial.')),
  'sn'            => array('text', 'textfield', t('Last Name'), 64, 64), //aka Surname
  'cn'            => array('text', 'textfield', t('Common Name'), 64, 64),
//  'displayName'   => array('text', 'textfield', t('Display Name'), 64, 64),
//  'mail'          => array('mailto', 'textfield', t('Email'), 64, 64, ''),
//  'description'   => array('text', 'textfield', t('Description'), 64, 64, ''),
//  'postOfficeBox' => array('text', 'textfield', t('P.O. Box'), 64, 64),
//  'streetaddress' => array('text', 'textfield', t('Address'), 64, 64),
//  'l'             => array('text', 'textfield', t('City'), 64, 64, t('Your city or other location')),
//  'st'            => array('text', 'textfield', t('State'), 4, 2, t('Your state or province')),
//  'postalCode'    => array('text', 'textfield', t('Postal Code'), 7, 5, t('Your Zip or Postal code')),
//  'co'            => array('text', 'textfield', t('Country'), 64, 64, t('Country or region')),
//  'company'       => array('text', 'textfield', t('Company'), 12, 12, 'Company or organization'),
//  'title'         => array('text', 'textfield', t('Title'), 12, 12),
//  'department'    => array('text', 'textfield', t('Department'), 64, 64),
//  'physicalDeliveryOfficeName' => array('text', 'textfield', t('Office'), 64, 64, ''),
//  'homephone'     => array('text', 'textfield', t('Home Phone'), 64, 64, ''),
//  'mobile'        => array('text', 'textfield', t('Mobile Phone'), 64, 64, ''),
//  'otherTelephone' => array('text', 'textfield', t('Other Phone'), 64, 64, ''),
//  'wWWHomePage'   => array('image', 'textfield', t('Web Page'), 64, 64, ''),
//  'url'           => array('link', 'textfield', t('Other web pages'), 64, 64, ''),
//  'roomNumber'    => array('text', 'textfield', t('Office Location'), 64, 64, ''),
//  'telephoneNumber' => array('text', 'textfield', t('Office Phone'), 64, 64, ''),
  'loginShell'            => array('text', 'textfield', t('Login shell'), 64, 64),

  // examples of displaying attributes in Select menus or TextArea boxes. Helpful for composing search queries
//  'gender'        => array('text', 'select', t('Gender'), array(0 => '-', 'm' => t('male'), 'f' => t('female')), 'Pick a gender', 0, 0),
//  'publickey'     => array('text', 'textarea', t('Public key'), 64, 4, '')
);

?>

ldap_integration/ldapdata.conf.php
<?php
// $Id$

// WARNING: if you change this, then you WANT to set attribute mapping
// to read-only. Otherwise, odd stuff WILL happen. Seriously!!!!!!
function ldapdata_attribute_filter(&$attributes) {
  // Uncomment this line to see what you receive as argument
  //msg_r($attributes);

  // Example:
  //     Say the attrribute emailAddress is stored like
  //   whatever#jdoe@example.com . You obviously will want
  //   Drupal to have it as jdoe@example.com, so you do this:
  //$attributes['emailAddress'][0] = preg_replace('/[^#]+#([^#])/', '$1', $attributes['emailAddress'][0]);

  return $attributes;
}


// Edit this list so that all attributes that you might wish to show
// are included. You may then enable/disable attributes in the Admin page.
//
// The left hand element indicates the LDAP attribute name.
//
// The 1st right hand elements in the array determine how the attribute and
// its value are displayed. The first element may be 'text' or 'url'.
$GLOBALS['ldap_attributes'] = array(
  'givenName'			=> array('text', 'textfield', t('First Name'), 64, 64),
  'sn'				=> array('text', 'textfield', t('Last Name'), 64, 64), //aka Surname
  'cn'            => array('text', 'textfield', t('Common Name'), 64, 64),
//  'homePage'      => array('url', 'textfield', t('Other web pages'), 64, 64),
  'mail'          => array('text', 'textfield', t('Email'), 64, 64)
);

?>


ldap_integration/ldapgroups.conf.php
<?php
// $Id$

// Interesting constants that admins would want to mess with

//   The module automatically decides names for the Drupal roles
// based in the names of the LDAP groups. For example:
//   - LDAP group: Admins => Drupal role: Admins
//   - LDAP group: ou=Underlings,dc=myorg,dc=mytld => Drupal role: Underlings
//   However, if this is not enough, this name mapping can be refined
// by altering this array. Some examples are given.


$GLOBALS['ldap_group_role_mappings'] = array(
  // LDAP group => Drupal role
  
  // make sure the last group->role mapping does NOT have a trailing comma (,)
  //'cn=admin,ou=Group,dc=example,dc=com' => 'IT'
  
);

// Note: Uncommenting this function will limit the groups -> roles conversion to ONLY those groups that are 
// specified in the function. 
/*
function ldapgroups_roles_filter($groups) { 
  global $ldap_group_role_mappings; 
  $roles = array(); 
  // this should take the roles array, pass it thru the filters and send a NEW set of roles back the filter 
  foreach ( $groups as $group ) { 
    foreach ($ldap_group_role_mappings as $approved_group => $approved_role) {
       // must strip spaces ?
       $group_stripped = preg_replace('/\s+/', '', $group);
       $approved_group_stripped = preg_replace('/\s+/', '', $approved_group);       
      if (strcasecmp($approved_group_stripped, $group_stripped) == 0) {
        // this role is specified -- grant
	   $roles[] = $approved_role;
      }
    }
  }
  return $roles;
}
*/

?>


ldap_integration/LDAPInterface.php
<?php
// $Id$

class LDAPInterface {

  function LDAPInterface() {
    $this->connection = NULL;
    //http://drupal.org/node/158671
    $this->server = NULL;
    $this->port = "389";
    $this->secretKey = NULL;
    $this->tls = false;
    $this->attr_filter = array('LDAPInterface', '__empty_attr_filter');
  }

  var $connection;
	var $server;
	var $port;
	var $tls;
	var $attr_filter;

  // This should be static, but that's not supported in PHP4
  function __empty_attr_filter($x) {
    return $x;
  }

  function setOption($option, $value) {
		switch($option) {
			case 'name':
				$this->name = $value;
				break;
			case 'server':
				$this->server = $value;
				break;
			case 'port':
				$this->port = $value;
				break;
			case 'tls':
				$this->tls = $value;
				break;
			case 'encrypted':
				$this->encrypted = $value;
				break;
			case 'user_attr':
				$this->user_attr = $value;
				break;
			case 'attr_filter':
				$this->attr_filter = $value;
				break;
			case 'basedn':
				$this->basedn = $value;
				break;
			case 'mail_attr':
			  $this->mail_attr = $value;
			  break;
		}
  }

  function getOption($option) {
		$ret = '';
		switch($option) {
			case 'version':
				$ret = -1;
				ldap_get_option($this->connection, LDAP_OPT_PROTOCOL_VERSION, $ret);
				break;
			case 'name':
				$ret = $this->name;
				break;
			case 'port':
				$ret = $this->port;
				break;
			case 'tls':
				$ret = $this->tls;
				break;
			case 'encrypted':
				$ret = $this->encrypted;
				break;
			case 'user_attr':
				$ret = $this->user_attr;
				break;
			case 'attr_filter':
				$ret = $this->attr_filter;
				break;
			case 'basedn':
				$ret = $this->basedn;
				break;
			case 'mail_attr':
			  $ret = $this->mail_attr;
			  break;
		}
		return $ret;
  }

  function connect($dn = '', $pass = '') {
    $ret = FALSE;
	// http://drupal.org/node/164049
    // If a connection already exists, it should be terminated
    $this->disconnect();

    if ($this->connectAndBind($dn, $pass)) {
      $ret = TRUE;
    }

    return $ret;
  }

  function initConnection() {
    if (!$con = ldap_connect($this->server, $this->port)) {
      watchdog('user', 'LDAP Connect failure to ' . $this->server . ':' . $this->port);
      return NULL;
    }

    $this->connection = $con;
    ldap_set_option($con, LDAP_OPT_PROTOCOL_VERSION, 3);
    ldap_set_option($con, LDAP_OPT_REFERRALS, 0);   
    // TLS encryption contributed by sfrancis@drupal.org
    if ($this->tls) {
      $vers = $this->getOption('version');
      if ($vers == -1) {
        watchdog('user', 'Could not get LDAP protocol version.');
      }

      if ($vers != 3) {
        watchdog('user', 'Could not start TLS, only supported by LDAP v3.');
      }
      else if (!function_exists('ldap_start_tls')) {
        watchdog('user', 'Could not start TLS. It does not seem to be supported by this PHP setup.');
      }
      else if (!ldap_start_tls($con)) {
        watchdog('user', t("Could not start TLS. (Error %errno: %error).", array('%errno' => ldap_errno($con), '%error' => ldap_error($con))));
      }
    }
  }

  function connectAndBind($dn = '', $pass = '') {
    $this->initConnection();

    $con = $this->connection;
    if (!$this->bind($dn, $pass)) {
      watchdog('user', t('LDAP Bind failure for user %user. Error %errno: %error', array('%user' => $dn,'%errno' => ldap_errno($con), '%error' => ldap_error($con))));
      return NULL;
    }

    return $con;
  }

  function bind($dn, $pass) {
    ob_start();
    set_error_handler(array('LDAPInterface', 'void_error_handler'));
    $ret = ldap_bind($this->connection, $dn, $pass);
    restore_error_handler();

    ob_end_clean();

    return $ret;
  }

  function disconnect() {
    if ($this->connection) {
      ldap_unbind($this->connection);
      $this->connection = NULL;
    }
  }

  function search($base_dn, $filter, $attributes = array()) {
    $ret = array();

    set_error_handler(array('LDAPInterface', 'void_error_handler'));
    $x = @ldap_search($this->connection, $base_dn, $filter, $attributes);
    restore_error_handler();

    if ($x && ldap_count_entries($this->connection, $x)) {
      $ret = ldap_get_entries($this->connection, $x);
    }
    return $ret;
  }

  // WARNING! WARNING! WARNING!
  // This function returns its entries with lowercase attribute names.
  // Don't blame me, blame PHP's own ldap_get_entries()
  function retrieveAttributes($dn) {
    set_error_handler(array('LDAPInterface', 'void_error_handler'));
    $result = ldap_read($this->connection, $dn, 'objectClass=*');
    $entries = ldap_get_entries($this->connection, $result);
    restore_error_handler();

    return call_user_func($this->attr_filter, $entries[0]);
  }

  function retrieveAttribute($dn, $attrname) {
    $entries = $this->retrieveAttributes($dn);
    return $entries[strtolower($attrname)][0];
  }

  function retrieveMultiAttribute($dn, $attrname) {
    $entries = $this->retrieveAttributes($dn);

    $result = array();
    $retrieved = $entries[strtolower($attrname)];
    $retrieved = $retrieved ? $retrieved : array();
    foreach ($retrieved as $key => $value) {
      if ($key !== 'count') {
        $result[] = $value;
      }
    }
    return $result;
  }

  function writeAttributes($dn, $attributes) {
    foreach ($attributes as $key => $cur_val) {
      if ($cur_val == '') {
        unset($attributes[$key]);
        $old_value = $this->retrieveAttribute($dn, $key);
        ldap_mod_del($this->connection, $dn, array($key => $old_value));
      }
      if (is_array ($cur_val)) {
        foreach ($cur_val as $mv_key => $mv_cur_val) {
          if ($mv_cur_val == '') {
            unset ($attributes[$key][$mv_key]);
          } else {
            $attributes[$key][$mv_key] = $mv_cur_val;
          }
        }
      }
    }

    ldap_modify($this->connection, $dn, $attributes);
  }

  // This should be static, but that's not supported in PHP4
  function void_error_handler($p1, $p2, $p3, $p4, $p5) {
    // Do nothing
  }
}

?>
